homeassistant/components/climate/generic_thermostat.py

```diff
@@ -302,65 +390,143 @@ def _async_update_temp(self, state):
             _LOGGER.error('Unable to update from sensor: %s', ex)
 
     @callback
-    def _async_control_heating(self):
+    def _async_control_thermostat(self):
         """Check if we need to turn heating on or off."""
-        if not self._active and None not in (self._cur_temp,
-                                             self._target_temp):
+        if not self._active and not self._dual_temp and \
+                None not in (self._cur_temp, self._target_temp):
             self._active = True
             _LOGGER.info('Obtained current and target temperature. '
                          'Generic thermostat active.')
+        if not self._active and self._dual_temp and None not in \
+                (self._cur_temp, self._target_temp_high,
+                 self._target_temp_low):
+            self._active = True
+            _LOGGER.info('Obtained current and target temperatures. '
+                         'Generic thermostat active.')
 
         if not self._active:
             return
 
-        if not self._enabled:
+        if self.current_operation == STATE_OFF:
             return
 
         if self.min_cycle_duration:
-            if self._is_device_active:
+            if self._is_cooling:
                 current_state = STATE_ON
             else:
                 current_state = STATE_OFF
-            long_enough = condition.state(
+            long_enough_cooling = condition.state(
+                self.hass, self.ac_entity_id, current_state,
+                self.min_cycle_duration)
+            if self._is_heating:
+                current_state = STATE_ON
+            else:
+                current_state = STATE_OFF
+            long_enough_heating = condition.state(
                 self.hass, self.heater_entity_id, current_state,
                 self.min_cycle_duration)
-            if not long_enough:
-                return
-
-        if self.ac_mode:
-            is_cooling = self._is_device_active
-            if is_cooling:
-                too_cold = self._target_temp - self._cur_temp >= \
-                    self._cold_tolerance
-                if too_cold:
-                    _LOGGER.info('Turning off AC %s', self.heater_entity_id)
-                    self._heater_turn_off()
+            if self.current_operation == STATE_COOL:
+                if not long_enough_cooling:
+                    return
+            if self.current_operation == STATE_HEAT:
+                if not long_enough_heating:
+                    return
+            if self.current_operation == STATE_AUTO:
+                if not long_enough_cooling or not long_enough_heating:
+                    return
+
+        if self.current_operation == STATE_COOL:
```

Comment:
Can't you do `if self.current_operation in (STATE_COOL, STATE_AUTO)` here, do the same for heating, and get rid of the block handling `STATE_AUTO` below (line 478)?

Reply:
I've modified the logic to use entities instead of states. It makes more sense as you can just do checks on each entity if it changed regardless of everything else.